README.md
DiskJockey
===============

[![Build Status](https://travis-ci.org/iancze/DiskJockey.svg?branch=master)](https://travis-ci.org/iancze/DiskJockey)

![Logo](logo.png)

Copyright Ian Czekala and collaborators 2014-16

email: iancze@gmail.com

This package derives dynamical masses for T Tauri stars using the Keplerian motion of their circumstellar disks, applied to radio interferometric data from the Atacama Large Millimeter Array (ALMA) and the Submillimeter Array (SMA). **If you use this code or a derivative of it in your research, you must cite [Czekala et al. 2015 ApJ, 806 154C](http://adsabs.harvard.edu/abs/2015ApJ...806..154C).**

See an explanation of [how dynamical mass measurements](http://iancze.github.io/dynamical/) work.

Papers published using DiskJockey:

* *A Disk-based Dynamical Constraint on the Mass of the Young Binary AK Sco*: [Czekala et al. 2015 ApJ, 806 154C](http://adsabs.harvard.edu/abs/2015ApJ...806..154C)
* *A Disk-based Dynamical Constraint on the Mass of the Young Binary DQ Tau* : [Czekala et al. 2016 ApJ, 818 156C](http://adsabs.harvard.edu/abs/2016ApJ...818..156C)

We can also be found in the [astrophysics source code library](http://ascl.net/1603.011) as well.

# Organization

DiskJockey is designed to forward model interferometric observations of protoplanetary disks, for the purpose of deriving a precise measurement of the central (sub-)stellar mass.

This package relies upon the excellent radiative synthesis package RADMC-3D to perform the radiative transfer of the disk model. Comprehensive documentation for RADMC-3D can be found [here](http://www.ita.uni-heidelberg.de/~dullemond/software/radmc-3d/). For package versions after v0.1.2, RADMC-3D is installed locally within the DiskJockey package. This means that you should be able to run all of the examples without needing to install RADMC-3D separately. If you already use RADMC-3D for other synthesis, this will leave your current installation alone, it's just that DiskJockey will use the version it installed for itself.

There are several stages to evaluating the best-fitting parameters of a disk structure model:

1. model specification
2. radiative synthesis
3. visibility interpolation
4. likelihood evaluation

I have organized the code in this package into source (in the `src` directory) and scripts (in the `scripts`) directory. After installation (see below), if you do

    using DiskJockey

You will be able to access any of the modules within `src`. These provide the base functionality for the package in common tasks, such as reading in images produced by RADMC-3D, Fourier transforms, and visibility interpolation. My hope is to make these components as general as possible so if you would like to extend this package to fit a novel type of disk, it will be easy to reuse many of the core functionality.

The scripts are not technically part of the Julia package (you cannot import them like the modules) but instead provide "driver scripts" that utilize the core modules to address a certain research question. For example, `DJ_initialize.jl` writes input files to disk, and `DJ_synthesize_model.jl` synthesizes a model using RADMC-3D. These are run from your system shell after adding them to your `PATH`

    $ DJ_initialize.jl && DJ_synthesize_model.jl

## Computational demand

Because spectral line datasets are large, and synthesizing models is computationally expensive, I have designed this package to work in a parallel environment. Therefore, please keep this architecture in mind when navigating the source code. Due to the computationally expensive nature of the radiative synthesis, fitting sizable datasets (e.g., SMA and ALMA) will **require a substantial amount of CPU cores to explore a posterior distribution in a reasonable timeframe**. For example, to fully explore the posterior for the AK Sco example dataset (to a density comparable to the plots in the ApJ paper) will require a **few days** of computing time on ~32 cores or more.

## Installation

### Julia

First, you should install the [Julia programming language](http://julialang.org/) on your machine. Instructions can be found [here](http://julialang.org/downloads/). I have found it exciting to install from source via the [github repo](https://github.com/JuliaLang/julia/), but it may be a bit quicker to use a pre-compiled binary. If you've successfully installed everything, you should be able to open up a Julia interpreter and type

    $ julia
                   _
       _       _ _(_)_     |  A fresh approach to technical computing
      (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
       _ _   _| |_  __ _   |  Type "?help" for help.
      | | | | | | |/ _` |  |
      | | |_| | | | (_| |  |  Version 0.4.2-pre+12 (2015-11-19 02:34 UTC)
     _/ |\__'_|_|_|\__'_|  |  Commit 9e598b6* (83 days old release-0.4)
    |__/                   |  x86_64-unknown-linux-gnu

    julia> println("Hello world")
    Hello world

Depending on how you choose to install Julia, you may need to take the additional step of adding the Julia executable to your `PATH`. [Here](https://en.wikibooks.org/wiki/Introducing_Julia/Getting_started#Running_directly_from_terminal) are some suggestions for OS X if you are experiencing difficulty.

### Fortran

Some of the packages that DiskJockey requires (e.g. [Dierckx.jl](https://github.com/kbarbary/Dierckx.jl)) require a Fortran compiler. On linux, try looking to make sure you have `gcc/gfortran` or something like it installed. For OS X, you can download these packages from [here](http://hpc.sourceforge.net/).

### DiskJockey

Next, we will install the `DiskJockey` package itself. Because this is not yet an official Julia package, for now, installation involves simply cloning the repository. First, open up a Julia prompt in the REPL, then type

    julia> Pkg.clone("https://github.com/iancze/DiskJockey.git")
    julia> Pkg.build("DiskJockey")

This process may take a few minutes as the relevant packages (including RADMC-3D) are downloaded from the web and installed. So far, I have only been able to extensively test this installation process on Linux machines. If you run into errors in this build process, please file an [issue](https://github.com/iancze/DiskJockey/issues) on the github repository so that we may try to fix this. If you already have RADMC-3D installed on your system, this process won't interfere with that installation, but `DiskJockey` will simply use the version of RADMC-3D it downloaded.

As mentioned previously, there are several "driver" command line scripts that are used to perform the actual mass fitting. To complete the installation, you should add these files to your system PATH. To figure out where the package is installed

    julia> Pkg.dir("DiskJockey")
    "/home/ian/.julia/DiskJockey"

Your PATH will vary. The scripts are located inside of the `scripts` directory, so if you are using bash or Z-shell, you will want to add the PATH that looks something like

    export PATH="/home/ian/.julia/DiskJockey/scripts:$PATH"

inside of your `.bashrc` or `.zshrc` file.

Some of the Python scripts also depend on imports from this directory, so you will also need to add

    export PYTHONPATH="/home/ian/.julia/DiskJockey/scripts:$PYTHONPATH"

inside of your `.bashrc` or `.zshrc` file. Finally,

    $ source ~/.zshrc

To check that you have properly added the scripts (and check that you installed the correct version), you can try in your system shell

    $ DJ_initialize.jl --version
    Your DiskJockey scripts are successfully linked.
    You are running DiskJockey 0.1.1
    Exiting

Due to the ongoing development of this package, it is easiest to keep your version current by updating off of the git master branch. For example,

    julia> Pkg.dir("DiskJockey")
    "/home/ian/.julia/DiskJockey

    # Exit Julia
    # Navigate to this directory
    $ cd /home/ian/.julia/DiskJockey
    $ git pull

For archival purposes, tagged release versions of this package are available [here](https://github.com/iancze/DiskJockey/releases).

### Python scripts

Lastly, some of the analysis scripts and IO routines require Python and several Python packages. I have only tested the scripts on Python 3.x, although they *may* work on Python 2.7. Please install the following packages via your own package manager or a distribution like anaconda:

* numpy
* scipy
* matplotlib
* h5py
* PyYAML
* Jupyter/IPython
* [corner.py](https://github.com/dfm/corner.py)

With the package successfully installed, see the documentation in the `docs/` folder on how to get started fitting a specific source, in particular the [cookbook](docs/cookbook.md).

If you'd like to run the test suite to make sure everything checks out, start Julia and run

    julia> Pkg.test("DiskJockey")

this may take about 10 minutes or so. If you catch any errors for your specific machine, please report them via an Issue on the github repository.

The code package is designed to interface with visibilities in the UVHDF5 format, described [here](https://github.com/Astrochem/UVHDF5). There are also scripts within this repository to convert to and from UVFITS and CASA measurement sets.

For those interested in the source code, the most important files to start browsing are

**model.jl**: Contains the specification of the parametric disk model, as well as the tools to write to disk the synthesis files RADMC-3D requires.

**image.jl**: Contains type definitions to read images produced by RADMC-3D, as well as convert from physical coordinates to sky coordinates.

**visibilities.jl**: Contains type definitions to hold the dataset and the model visibilities. Additionally contains functions to apply phase shifts to the visibilities corresponding to shifts in the image plane. Also contains functions to FFT images to the visibility plane.

**gridding.jl**: Contains the prolate-spheroidal wave function definitions from [Schwab 1984](http://adsabs.harvard.edu/abs/1984iimp.conf..333S), used when doing the visibility interpolations.

**venus.jl**: This implementation uses the Ensemble Sampler (a Julia port of Dan Foreman-Mackey's `emcee` python package) to sample the posterior distribution using parallelized walkers.


docs/changelog.md
# Changelog

The following are the changes that have been implemented since the previous version. All versions are tagged as releases and available as a clone off of the master branch.

# Version 0.1.4

Upgraded to Julia version 0.5. To upgrade to this version of DiskJockey, you will first need to upgrade your own Julia distribution to 0.5 as well. I apologize for the frequent upgrades of Julia, but since Julia is itself a fast-developing language, it makes the most sense to just bite the bullet and upgrade every time the programming language upgrades as well. Once Julia reaches v1.0 (sometime in 2017, perhaps), these changes should become less frequent.

After you install the new version of julia, make sure that it actually loads as

    $ julia
                   _
       _       _ _(_)_     |  A fresh approach to technical computing
      (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
       _ _   _| |_  __ _   |  Type "?help" for help.
      | | | | | | |/ _` |  |
      | | |_| | | | (_| |  |  Version 0.5.1-pre+31 (2016-11-17 17:50 UTC)
     _/ |\__'_|_|_|\__'_|  |  Commit 6a1e339* (7 days old release-0.5)
    |__/                   |  x86_64-pc-linux-gnu

    julia>


And then you will need to reinstall DiskJockey in this new version, which should automatically pull down all of the other packages.

    julia> Pkg.clone("https://github.com/iancze/DiskJockey.git")

If you no longer plan on using Julia v0.4 or any of the packages, I would recommend deleting it from your system so as not to cause any PATH conflicts. Speaking of which, don't forget to update your PATH to point to the new v0.5 version scripts.

## Renaming

The script `max_baseline.jl` is now `DJ_max_baseline.jl`. `plot_walkers.py` is now `DJ_plot_walkers.py`.

## DJ_plot_walkers.py

Now has proper labeling for models that fix parameters.

## DJ_verify_run.jl

A new script for checking that you've specified everything properly before launching a MCMC run.

# Version 0.1.3

## Fix parameters

Now there is a flexible way to fix or free parameters in a fit.

To do this requires new config files (which you may need to copy over from `assets/`) that include a `params_fixed` field, where you list the parameters to fix. Then, it may also require modifying the `InitializeWalkers.ipynb` notebooks to incorporate more or fewer parameters than are currently available.


## Cavity model

Now supports a cavity model, with parameters for the size of the cavity (`r_cav`), as well as the steepness (`gamma_cav`).

## Truncated model

Now includes a model that has a variable outer exponential taper, allowing it to be steeper or shallower than the traditional (2 - gamma).

## Change to Sigma_c

Instead of fitting with the parameter total gas mass, we now fit with the surface density normalization at the critical radius, `Sigma_c`. Note that this only truly the normalization constant for the `standard`, `truncated`, and `vertical` models. The reason for the switch from `M_gas` to `Sigma_c` is that for more complicated models, there is no analytic formula for the total gas mass, meaning that a numerical integral would be needed for each model evaluation. It is simpler and more accurate to sample in `Sigma_c` and then later convert the samples to `M_gas` if desired.

## FITS export

Thanks to Jane Huang (@j6626), you can now export a RADMC image to a FITS file via the script `DJ_image_to_FITS.py`. You can then inspect these FITS files with `ds9`, or use them in CASA `simobserve`.

## spectrum.png

The plotting routine now outputs integrated line flux, to be used as a sanity check against the measured value from a dataset.

## Half-pixel offset

Has now been added to the code. This is a result from that RADMC synthesizes an image centered on `(0,0)`, while the FFT routine expects the image to be centered on the middle of the central pixel. The only change you should see is a small, half-pixel sized offset in `mu_RA` and `mu_DEC`.

## Visualizing the tau=1 surface

This functionality originally provided in RADMC is now exposed via DiskJockey to visualize the tau = 1 (or other value) surface.

  $ DJ_initialize.jl && tausurf_model.jl && plot_tausurf.jl

This also introduces the `TausurfImage` type in `src/image.jl`. This is useful to see whether the tau=1 surface is in front of or behind the midplane of the disk, projected on the sky.

It will be more useful to plot this in 3D, however. ParaView export is hopefully coming in a future version.

## Vertical temperature gradient

We now include support for fitting a vertical temperature gradient, via the `vertical` model type. We follow the parameterization in [Williams and Best 14](https://ui.adsabs.harvard.edu/#abs/2014ApJ...788...59W/abstract).

## model.convert_vector

This routine is to simplify to ingesting *kwargs*, enabling easier selection of model types and parameters.

## User-defined priors

As an experimental feature, it is now possible to assign user defined priors. This will require you to write some Julia code, but shouldn't be that difficult. An example is in `assets/prior.jl`, which will be read automatically from the current working directory. The main idea is to make it easier to assign disk-specific priors (i.e., constrain the disk gas mass to a specific value, or disk position, etc...).

## Makefiles

Now, we include a makefile that should be copied to each source directory. In a typical analysis workflow, there isn't need to regenerate each intermediate product repeatedly. Details are in the cookbook.

## Script renaming

To prevent these scripts from cluttering a users namespace when adding to their `PATH`, we've prefixed most things with `DJ_`. Moreover, for most tasks, the users will not need to run these tasks directly, but will just use the Makefile.

## model.Grid

`Grid` type initialization now relies upon a dictionary.

# Version 0.1.2

The package is now named DiskJockey (previously JudithExcalibur). Check out the new logo!

## Installation


hdfj.py
#!/usr/bin/env python

import argparse

parser = argparse.ArgumentParser(description="Measure statistics across multiple chains.")
parser.add_argument("--glob", help="Do something on this glob. Must be given as a quoted expression to avoid shell expansion.")

# parser.add_argument("--dir", action="store_true", help="Concatenate all of the flatchains stored
# within run* folders in the current directory. Designed to collate runs from a JobArray.")
parser.add_argument("--outdir", default="mcmcplot", help="Output directory to contain all plots.")
parser.add_argument("--output", default="combined.hdf5", help="Output HDF5 file.")

parser.add_argument("--files", nargs="+", help="The HDF5 or CSV files containing the MCMC samples, separated by whitespace.")

parser.add_argument("--burn", type=int, default=0, help="How many samples to discard from the beginning of the chain "
                                                        "for burn in.")
parser.add_argument("--thin", type=int, default=1, help="Thin the chain by this factor. E.g., --thin 100 will take every 100th sample.")
parser.add_argument("--range", nargs=2, help="start and end ranges for chain plot, separated by WHITESPACE")

parser.add_argument("--cat", action="store_true", help="Concatenate the list of samples.")
parser.add_argument("--chain", action="store_true", help="Make a plot of the position of the chains.")
parser.add_argument("-t", "--triangle", action="store_true", help="Make a triangle (staircase) plot of the parameters.")
parser.add_argument("--params", nargs="*", default="all", help="A list of which parameters to plot,                                                                 separated by WHITESPACE. Default is to plot all.")

parser.add_argument("--gelman", action="store_true", help="Compute the Gelman-Rubin convergence statistics.")

parser.add_argument("--acor", action="store_true", help="Calculate the autocorrelation of the chain")
parser.add_argument("--acor-window", type=int, default=50, help="window to compute acor with")

parser.add_argument("--cov", action="store_true", help="Estimate the covariance between two parameters.")
parser.add_argument("--ndim", type=int, help="How many dimensions to use for estimating the 'optimal jump'.")
parser.add_argument("--paper", action="store_true", help="Change the figure plotting options appropriate for the paper.")
parser.add_argument("--save_npy", action="store_true", help="Save only (Mass, distance, inclination) parameters to a separate numpy file.")

args = parser.parse_args()

import os
import numpy as np
import h5py
from astropy.table import Table
from astropy.io import ascii
import sys
import csv

#Check to see if outdir exists.
if not os.path.exists(args.outdir):
    os.makedirs(args.outdir)

args.outdir += "/"



def h5read(fname, burn=0, thin=1):
    '''
    Read the flatchain from the HDF5 file and return it.
    '''
    fid = h5py.File(fname, "r")
    assert burn < fid["samples"].shape[0]
    print("{} burning by {} and thinning by {}".format(fname, burn, thin))
    flatchain = fid["samples"][burn::thin]

    fid.close()

    return flatchain

def csvread(fname, burn=0, thin=1):
    '''
    Read the flatchain from a CSV file and return it.
    '''
    flatchain = np.genfromtxt(fname, skip_header=1, dtype=float, delimiter=",")[burn::thin]

    return flatchain

def gelman_rubin(samplelist):
    '''
    Given a list of flatchains from separate runs (that already have burn in cut
    and have been trimmed, if desired), compute the Gelman-Rubin statistics in
    Bayesian Data Analysis 3, pg 284. If you want to compute this for fewer
    parameters, then slice the list before feeding it in.
    '''

    full_iterations = len(samplelist[0])
    assert full_iterations % 2 == 0, "Number of iterations must be even. Try cutting off a different number of burn in samples."
    shape = samplelist[0].shape
    #make sure all the chains have the same number of iterations
    for flatchain in samplelist:
        assert len(flatchain) == full_iterations, "Not all chains have the same number of iterations!"
        assert flatchain.shape == shape, "Not all flatchains have the same shape!"

    #make sure all chains have the same number of parameters.

    #Following Gelman,
    # n = length of split chains
    # i = index of iteration in chain
    # m = number of split chains
    # j = index of which chain
    n = full_iterations//2
    m = 2 * len(samplelist)
    nparams = samplelist[0].shape[-1] #the trailing dimension of a flatchain

    #Block the chains up into a 3D array
    chains = np.empty((n, m, nparams))
    for k, flatchain in enumerate(samplelist):
        chains[:,2*k,:] = flatchain[:n]  #first half of chain
        chains[:,2*k + 1,:] = flatchain[n:] #second half of chain

    #Now compute statistics
    #average value of each chain
    avg_phi_j = np.mean(chains, axis=0, dtype="f8") #average over iterations, now a (m, nparams) array
    #average value of all chains
    avg_phi = np.mean(chains, axis=(0,1), dtype="f8") #average over iterations and chains, now a (nparams,) array

    B = n/(m - 1.0) * np.sum((avg_phi_j - avg_phi)**2, axis=0, dtype="f8") #now a (nparams,) array

    s2j = 1./(n - 1.) * np.sum((chains - avg_phi_j)**2, axis=0, dtype="f8") #now a (m, nparams) array

    W = 1./m * np.sum(s2j, axis=0, dtype="f8") #now a (nparams,) arary

    var_hat = (n - 1.)/n * W + B/n #still a (nparams,) array
    std_hat = np.sqrt(var_hat)

    R_hat = np.sqrt(var_hat/W) #still a (nparams,) array


    data = Table({   "Value": avg_phi,
                     "Uncertainty": std_hat},
                 names=["Value", "Uncertainty"])

    print(data)

    ascii.write(data, sys.stdout, Writer = ascii.Latex, formats={"Value":"%0.3f", "Uncertainty":"%0.3f"}) #

    #print("Average parameter value: {}".format(avg_phi))
    #print("std_hat: {}".format(np.sqrt(var_hat)))
    print("R_hat: {}".format(R_hat))

    if np.any(R_hat >= 1.1):
        print("You might consider running the chain for longer. Not all R_hats are less than 1.1.")


def plot(flatchain, base=args.outdir, format=".png"):
    '''
    Make a triangle plot
    '''

    import triangle

    labels = [r"$M_\ast\quad [M_\odot]$", r"$r_c$ [AU]", r"$T_{10}$ [K]",
    r"$q$", r"$\log M_\textrm{CO} \quad \log [M_\oplus]$",  r"$\xi$ [km/s]",
    r"$d$ [pc]",
    r"$i_d \quad [{}^\circ]$", r"PA $[{}^\circ]$", r"$v_r$ [km/s]",
    r"$\mu_\alpha$ ['']", r"$\mu_\delta$ ['']"]
    figure = triangle.corner(flatchain, quantiles=[0.16, 0.5, 0.84],
        plot_contours=True, plot_datapoints=False, labels=labels, show_titles=True)
    figure.savefig(base + "triangle" + format)

def paper_plot(flatchain, base=args.outdir, format=".pdf"):
    '''
    Make a triangle plot of just M vs i
    '''

    import matplotlib
    matplotlib.rc("font", size=8)
    matplotlib.rc("lines", linewidth=0.5)
    matplotlib.rc("axes", linewidth=0.8)
    matplotlib.rc("patch", linewidth=0.7)
    import matplotlib.pyplot as plt
    #matplotlib.rc("axes", labelpad=10)
    from matplotlib.ticker import FormatStrFormatter as FSF
    from matplotlib.ticker import MaxNLocator
    import triangle

    labels = [r"$M_\ast\enskip [M_\odot]$", r"$i_d \enskip [{}^\circ]$"]
    #r"$r_c$ [AU]", r"$T_{10}$ [K]", r"$q$", r"$\log M_\textrm{CO} \enskip [\log M_\oplus]$",
    #r"$\xi$ [km/s]"]
    inds = np.array([0, 6, ]) #1, 2, 3, 4, 5])

    K = len(labels)
    fig, axes = plt.subplots(K, K, figsize=(3., 2.5))

    figure = triangle.corner(flatchain[:, inds], plot_contours=True,
    plot_datapoints=False, labels=labels, show_titles=False,
        fig=fig)

    for ax in axes[:, 0]:
        ax.yaxis.set_label_coords(-0.4, 0.5)
    for ax in axes[-1, :]:
        ax.xaxis.set_label_coords(0.5, -0.4)

    figure.subplots_adjust(left=0.2, right=0.8, top=0.95, bottom=0.2)

    figure.savefig(base + "ptriangle" + format)


def plot_walkers(flatchain, base=args.outdir, start=0, end=-1, labels=None):
    import matplotlib.pyplot as plt
    from matplotlib.ticker import MaxNLocator
    # majorLocator = MaxNLocator(nbins=4)
    ndim = len(flatchain[0, :])
    sample_num = np.arange(len(flatchain[:,0]))
    sample_num = sample_num[start:end]
    samples = flatchain[start:end]
    plt.rc("ytick", labelsize="x-small")

    # if lnprobs is not None:
    #     fig, ax = plt.subplots(nrows=ndim + 1, sharex=True)
    #     ax[0].plot(sample_num, lnprobs[start:end])
    #     ax[0].set_ylabel("lnprob")
    #     for i in range(0, ndim):
    #         ax[i+1].plot(sample_num, samples[:,i])
    #         ax[i+1].yaxis.set_major_locator(MaxNLocator(nbins=6, prune="both"))
    #         if labels is not None:
    #             ax[i+1].set_ylabel(labels[i])


    fig, ax = plt.subplots(nrows=ndim, sharex=True)
    for i in range(0, ndim):
        ax[i].plot(sample_num, samples[:,i])
        ax[i].yaxis.set_major_locator(MaxNLocator(nbins=6, prune="both"))
        if labels is not None:
            ax[i].set_ylabel(labels[i])

    ax[-1].set_xlabel("Sample number")
    fig.subplots_adjust(hspace=0)
    fig.savefig(base + "walkers.png")
    plt.close(fig)

def estimate_covariance(flatchain):
    print(flatchain.shape)
    print(np.average(flatchain.astype("f8"), axis=0)[-3])
    print(np.mean(flatchain, axis=0, dtype=np.float64)[-3])
    for var in flatchain.T:
        fmtstr = "Median: {:.4f}; Std: {:.4f}; 16th:{:.4f}; 84th:{:.4f}"
        print(fmtstr.format(np.median(var), np.std(var), np.percentile(var, 16), np.percentile(var, 84)))

    if args.ndim:
        d = args.ndim
    else:
        d = flatchain.shape[1]

    import matplotlib.pyplot as plt

    #print("Parameters {}".format(flatchain.param_tuple))
    #samples = flatchain.samples
    cov = np.cov(flatchain, rowvar=0)

    #Now try correlation coefficient
    cor = np.corrcoef(flatchain, rowvar=0)
    print("Correlation coefficient")
    print(cor)

    # Make a plot of correlation coefficient.

    fig, ax = plt.subplots(figsize=(0.5 * d, 0.5 * d), nrows=1, ncols=1)
    ext = (0.5, d + 0.5, 0.5, d + 0.5)
    ax.imshow(cor, origin="upper", vmin=-1, vmax=1, cmap="bwr", interpolation="none", extent=ext)
    fig.savefig("cor_coefficient.png")

    print("'Optimal' jumps with covariance (units squared)")

    opt_jump = 2.38**2/d * cov
    # opt_jump = 1.7**2/d * cov # gives about ??
    print(opt_jump)

    print("Standard deviation")
    std_dev = np.sqrt(np.diag(cov))
    print(std_dev)

    print("'Optimal' jumps")
    if args.ndim:
        d = args.ndim
    else:
        d = flatchain.shape[1]
    print(2.38/np.sqrt(d) * std_dev)

    np.save("mcmcplot/opt_jump.npy", opt_jump)



def cat_list(file, flatchainList):
    '''
    Given a list of flatchains, concatenate all of these and write them to a
    single HDF5 file.
    '''
    #Write this out to the new file
    print("Opening", file)
    hdf5 = h5py.File(file, "w")

    cat = np.concatenate(flatchainList, axis=0)

    # id = flatchainList[0].id
    # param_tuple = flatchainList[0].param_tuple

    dset = hdf5.create_dataset("samples", cat.shape, compression='gzip',
        compression_opts=9)
    dset[:] = cat
    # dset.attrs["parameters"] = "{}".format(param_tuple)

    hdf5.close()

#Now that all of the structures have been declared, do the initialization stuff.

if args.glob:
    from glob import glob
    files = glob(args.glob)
elif args.files:
    files = args.files
else:
    import sys
    sys.exit("Must specify either --glob or --files")

#Because we are impatient and want to compute statistics before all the jobs are finished,
# there may be some directories that do not have a flatchains.hdf5 file
flatchainList = []
for file in files:
    try:
        # If we've specified HDF5, use h5read
        # If we've specified csv, use csvread
        root, ext = os.path.splitext(file)
        if ext == ".hdf5":
            flatchainList.append(h5read(file, args.burn, args.thin))
        elif ext == ".csv":
            flatchainList.append(csvread(file, args.burn, args.thin))
    except OSError as e:
        print("{} does not exist, skipping. Or error {}".format(file, e))

print("Using a total of {} flatchains".format(len(flatchainList)))

if args.cat:
    assert len(flatchainList) > 1, "If concatenating samples, must provide more than one flatchain"
    cat_list(args.output, flatchainList)

# Assume that if we are using either args.chain, or args.triangle, we are only suppling one
# chain, since these commands don't make sense otherwise.
if args.chain:
    assert len(flatchainList) == 1, "If plotting Markov Chain, only specify one flatchain"
    plot_walkers(flatchainList[0])

if args.triangle:
    assert len(flatchainList) == 1, "If making Triangle, only specify one flatchain"
    plot(flatchainList[0])

if args.paper:
    assert len(flatchainList) == 1, "If making Triangle, only specify one flatchain"
    paper_plot(flatchainList[0])

if args.cov:
    assert len(flatchainList) == 1, "If estimating covariance, only specify one flatchain"
    estimate_covariance(flatchainList[0])

if args.gelman:
    assert len(flatchainList) > 1, "If running Gelman-Rubin test, must provide more than one flatchain"
    gelman_rubin(flatchainList)

if args.save_npy:
    assert len(flatchainList) == 1, "If plotting Markov Chain, only specify one flatchain"
    flatchain = flatchainList[0]

    np.save("mdi.npy", flatchain[:,np.array([0, 6, 7])])
